tests/posix/common/src/pthread.c

```diff
@@ -590,11 +590,11 @@ ZTEST(posix_apis, test_pthread_descriptor_leak)
 
 	zassert_ok(pthread_attr_init(&attr));
 	zassert_ok(pthread_attr_setstack(&attr, &stack_e[0][0], STACKS));
-
 	/* If we are leaking descriptors, then this loop will never complete */
 	for (size_t i = 0; i < CONFIG_MAX_PTHREAD_COUNT * 2; ++i) {
 		zassert_ok(pthread_create(&pthread1, &attr, create_thread1, NULL),
 			   "unable to create thread %zu", i);
 		zassert_ok(pthread_join(pthread1, &unused), "unable to join thread %zu", i);
+		usleep(USEC_PER_MSEC * 500U);
```

Comment:
This should not be necessary. Adding arbitrary delays is a slippery-slope way to work around synchronization issues. I would prefer to see a root cause first.

Reply:
This is never found failure if we do not switch to none tickless mode, and I find it may related to thread protection.
by default the  CONFIG_MAX_PTHREAD_MUTEX_COUNT=5,  CONFIG_MAX_PTHREAD_COND_COUNT=5,
and pthread_join will depend on those pthread_mutex and pthread_cond, and in the  test, it seems quickly exhaust them

Reply:
> Okay, had to read up on posix threads... The assumption is that the pthread_join() doesn't return until the thread is terminated. But posix is also relying on the underlying RTOS to clean up the thread. Do we have a case here where the pthread_join() is waiting on the pthread->state_cond variable as the indicator that the thread has exited... which it did, but the kernel hasn't yet cleaned up the thread?

Quite possible.

> pthread_exit() at the does the pthread_cond_broadcast() on the state change of the state_cond variable then at the end of the routine it destroys and does the k_thread_abort(). Does the pthread_cond_broadcast() allow the scheduler to redo who is running and thus allows the test thread to run again? Is there a change that can be done to ensure that pthread_join() doesn't actually return until the zephyr kernel has cleaned up the thread?

To me that sounds like change that is necessary here. If it can be isolated / reproduced reliably, which @hakehuang has done, and then we have a change such that the repro recipe results in a successful test, I think that would be sufficient to close the bug.

I suspect this issue has likely been present for A Long Time (TM). I doubt that it's a regression. I think that this new test just uncovered something that has always been there.

Reply:
I'm not super familiar with the internals of the scheduler/kernel/threads but pthread_join() is waiting on the condition variable that is signaled when the pthread is shut down. The actual thread function is wrapped by zephyr_thread_wrapper() which is calling pthread_exit() after the original pthread function has exited. The pthread_exit() resets variables around the pthread before signalling the condition variable that pthread_join() is waiting on. Then the zephyr thread running this exits. 

I supposed there could be a timing here where the scheduler gets to run allowing the main test thread to continue in its loop to create another pthread. In this case though, it would try to regenerate the same thread that hasn't quiet exited yet because the state variable is telling pthread system that it is available. At least this is what @hakehuang results are indicating:

`ASSERTION FAIL [0] @ WEST_TOPDIR/zephyr/kernel/sched.c:1782
	aborted _current back from dead
`

What if we change pthread_exit() to move the pthread_cond_broadcast() in line 395 to before line 393:

`
	if (self->state == PTHREAD_JOINABLE) {
		self->state = PTHREAD_EXITED;
		self->retval = retval;
		pthread_cond_broadcast(&self->state_cond);
	} else {
`

Changed to:

`	if (self->state == PTHREAD_JOINABLE) {
		pthread_cond_broadcast(&self->state_cond);
		self->state = PTHREAD_EXITED;
		self->retval = retval;
	} else {
`

The would prevent pthread_create() from trying to allocate that particular pthread if the scheduler lets the test code run again because the state would not be PTHREAD_EXITED.

p.s. I'm not even able to reproduce this though. @hakehuang, can you provide me guidance on how to reproduce?

Reply:
@dleach02 , Thanks a lot for this analysis.
if remove the delay, and move the pthread_cond_broadcast() in line 395 to before line 393 like below
'''
diff --git a/lib/posix/pthread.c b/lib/posix/pthread.c
index d561bd49e8..6486581bdb 100644
--- a/lib/posix/pthread.c
+++ b/lib/posix/pthread.c
@@ -390,9 +390,9 @@ void pthread_exit(void *retval)

        pthread_mutex_lock(&self->state_lock);
        if (self->state == PTHREAD_JOINABLE) {
+               pthread_cond_broadcast(&self->state_cond);
                self->state = PTHREAD_EXITED;
                self->retval = retval;
-               pthread_cond_broadcast(&self->state_cond);
        } else {
                self->state = PTHREAD_TERMINATED;
        }

'''
it still fails as below
'''
    Assertion failed at WEST_TOPDIR/zephyr/tests/posix/common/src/pthread.c:597: posix_apis_test_pthread_descriptor_leak: (pthread_join(pthread1, &unused) is non-zero)
unable to join thread 9
 FAIL - test_pthread_descriptor_leak in 0.011 seconds
===================================================================
START - test_sleep
ASSERTION FAIL [0] @ WEST_TOPDIR/zephyr/kernel/sched.c:1782
        aborted _current back from dead
E: r0/a1:  0x00000004  r1/a2:  0x000006f6  r2/a3:  0xff00002c
E: r3/a4:  0x00008609 r12/ip:  0x000242e0 r14/lr:  0x0000ae5d
E:  xpsr:  0x0000013f
E: Faulting instruction address (r15/pc): 0x0000c288
E: >>> ZEPHYR FATAL ERROR 4: Kernel panic on CPU 0
E: Current thread: 0x20038 (unknown)
E: Halting system
'''
the test command line is
'''
west build -b qemu_cortex_r5
west build -t run 
'''

Reply:
> with something like `CONFIG_SPIN_VALIDATE=y`? Enabling debug logging at the kernel level?

no help. same failure, and for qemu seems there no any kernel dump no matter I enable
`
CONFIG_DEBUG_COREDUMP=y
CONFIG_LOG=y
`

Reply:
I believe the root problem has to do with the pthread->state and the states that the thread transitions to when combined with the termination of the thread and call to k_thread_abort(). The k_thread_abort() will allow some rescheduling of the threads which allows an "active" or waiting thread to run before returning to the k_thread_abort(). But because the pthread structures are an array and contain the k_thread object, the test code gets to run again and will allocate the same pthread object again. This will cause the assert in the k_thread_abort() to occur because it can detect that the thread object has been reactivated. 

So it seems like the code needs to put the pthread object into a state that prevents any allocation until it is fully cleaned up.

But the states the pthread can go through are DETACTED, JOINABLE, TERMINATED, and EXITED. At initialization time, the objects are initialized to EXITED. To allocate one, it needs to be in TERMINATED or EXITED. The pthread_join() function is looking for the thread state of EXITED so it can retrieve the exit code.

I thought about maybe adding a state: PTHREAD_AVAILABLE and use that for the pthread_create() but that would require the pthread_join() to transition the pthread to AVAILABLE. I think, for the Zephyr implementation, that might be okay because we don't seem to allow multiple threads to "join" the pthread.. if we did then we would have to do larger logic changes from what I can see. But even if we do this, I'm not sure if the thread run dynamics won't still cause this to occur. 

Maybe we need to put the pthread object into "exiting" and have the idle thread/worker task wake up periodically and transition the pthread to a new "FREE" state and change the allocate logic to look for TERMINATED or FREE?

Reply:
True - I can't remember the exact name for that but inserting a transitional or "fake" state to simplify a model is usually a good solution.

I wish I could reproduce this on a board with a debugger. Can you recommend an entry level NXP board available that I could use to do that?

Reply:
@cfriedt , this issue only happens on qemu and renode simualtor.

Reply:
Can you update the issue with another repro command?

Reply:
below steps will recreate this issue

```
west build -b qemu_cortex_r5
west build -t run 
```

The problem is that if I use the `west build -t debugserver`, and then use gdb to connect it halts untill I quit gdb.

Reply:
I will add a new tickless configuration to check this issue